Azure token refresh

codablock opened the pull request:
## Description

This is a re-submit of https://github.com/oauth2-proxy/oauth2-proxy/pull/278 after rebasing on the current master and applying necessary fixes needed due to changes in master since the original PR was created.

2 Additional changes were needed to fix issues we encountered with refresh tokens:
1. ~091727b09c49c19744e4ef46c21ca6a1c35d1a28~: This ensures that a refresh is performed early enough and that there is no chance of not refreshing and then running into an expired token a second later.
2. ~7505b020684d2f46bcc2afd01b926dca7880b10b: Tokens redeemed through a refresh token do not contain email addresses anymore. This behavior was quite strange and I can't really say what is going on. I figured out the easiest and still safe solution would be to allow empty emails when `--email-domain=*` is used.~ EDIT: This is obsolete due to https://github.com/oauth2-proxy/oauth2-proxy/pull/770

## Motivation and Context

This PR implements Azure token refreshs, as requested in https://github.com/oauth2-proxy/oauth2-proxy/issues/260

## How Has This Been Tested?

We have this code (based on a [master](cfa1160) version shortly after 6.0.0 was released) running in production since a few weeks. The original [PR](https://github.com/oauth2-proxy/oauth2-proxy/pull/278) also contained test cases for the new functionality, these tests are also present in this PR.

## Checklist:

<!--- Go over all the following points, and put an `x` in all the boxes that apply. -->
<!--- If you're unsure about any of these, don't hesitate to ask. We're here to help! -->

- [ ] My change requires a change to the documentation or CHANGELOG.
- [x] I have updated the documentation/CHANGELOG accordingly.
- [x] I have created a feature (non-master) branch for my PR.

JoeSainsburys commented:
Did you ever get this working? I've tried using this branch and I've tried implementing it in the OIDC provider and I can't get an identity_token. [Here ](https://github.com/MicrosoftDocs/azure-docs/issues/34799#issuecomment-510983463) MS are pretty explicit in saying that you can't refresh identity tokens like this.

codablock commented:
@JoeSainsburys You have to use access tokens. At the same time, the default access tokens returned by Azure AD are for graph.microsoft.com, which probably not what you actually want (these tokens are also hard to verify due to some internal use of nonces). You must add a custom scope to the app registration in Azure and then use the scope id (without the `api://` part, so really only the ID) in `--resource` and the scope name as `--scope`. After that you'll get refreshable access tokens for your own scope.

codablock commented:
@JoeSainsburys it's also important to have https://github.com/oauth2-proxy/oauth2-proxy/pull/753 merged as well when you want to try it out, as otherwise you'll still get the graph.microsoft.com access tokens even if you specify --resource.

JoeSainsburys commented:
Ahhh, thank you! I was thinking it had to be something to do with it either being set up as a B2C app or actually using an access token. I'll give it a go, thanks.

JoeSainsburys commented:
I think I've got it working to the point where it gets a refreshed access token, but it falls down when attempting to query the /me endpoint. It says that the audience is wrong (which makes sense because we have changed it from graph.microsoft.com to our app). I think that this is probably an issue with my permissions setup. I've not got the Directory.AccessAsUser.All permission, realistically I'm probably not going to get this (our tenant has ~200,000 users and that sounds like delegated god mode.) Would you expect to see this error if I was missing that permission?

codablock commented:
@JoeSainsburys Unfortunately I can't help with that issue, as we're not using graph.microsoft.com in our applications. We solely use the access tokens for authentication and then use internal apis/resources to gather information about the logged in user.

If you only need access to graph.microsoft.com and do not plan on using the token to do your own authentication/authorization, it's probably ok to just stick with graph.microsoft.com and but use access instead of id tokens instead. This then means you can't really verify the access tokens by yourself, but you can use the token to access graph.microsoft.com.

Btw, I'm not an oauth/oidc/AD specialist and my knowledge is very limited in this regard, so please take my works with a grain of salt, especially as this is security related.

codablock commented:
https://github.com/oauth2-proxy/oauth2-proxy/commit/7505b020684d2f46bcc2afd01b926dca7880b10b will become obsolete after #770 is merged, so I'm going to remote that commit from here.

NickMeves commented on providers/azure.go:
```diff
@@ -151,6 +157,60 @@ func (p *AzureProvider) Redeem(ctx context.Context, redirectURL, code string) (s
+// RefreshSessionIfNeeded checks if the session has expired and uses the
+// RefreshToken to fetch a new ID token if required
+func (p *AzureProvider) RefreshSessionIfNeeded(ctx context.Context, s *sessions.SessionState) (bool, error) {
+	if s == nil || s.ExpiresOn.Add(-time.Minute*10).After(time.Now()) || s.RefreshToken == "" {
```
Let's keep this aligned with the refresh timing logic in other providers:

```
	if s == nil || (s.ExpiresOn != nil && s.ExpiresOn.After(time.Now())) || s.RefreshToken == "" {
		return false, nil
	}
```

Refreshing sessions is an area we've identified corner case bugs around (refresh request stampedes getting multiple new access tokens, edge of time expiration/refresh token issues). We'd like to handle them better globally (I think to solve your issue, there's better logic that can surround how the `--cookie-refresh` timer logic in general if I recall).

Will this also lead to odd behavior for users that specific an access token TTL shorter than 10 minutes?

codablock replied:
Hmm yeah I assume this will do strange stuff when shorter TTLs are used. I'll remove the commit from this PR now and count on a global fix in the future. When I implemented this fix I was actually hunting down another bug, but I did not know that at that time and tried to fix it this way, so it should be ok to remove it again.

NickMeves commented on providers/azure.go:
```diff
@@ -151,6 +157,60 @@ func (p *AzureProvider) Redeem(ctx context.Context, redirectURL, code string) (s
+	return true, nil
+}
+
+func (p *AzureProvider) redeemRefreshToken(ctx context.Context, s *sessions.SessionState) (err error) {
```
NIT: We're trying to move away from named return parameters in new code, but I see this convention is used all over this provider -- so we can stick with it if you prefer.

JoelSpeed replied:
I'd prefer if we don't add named returns even if they are common across the provider, can we update this one?

NickMeves commented on providers/azure.go:
```diff
@@ -74,6 +75,9 @@ func NewAzureProvider(p *ProviderData) *AzureProvider {
 	if p.ProtectedResource == nil || p.ProtectedResource.String() == "" {
 		p.ProtectedResource = azureDefaultProtectResourceURL
 	}
+	if p.ValidateURL == nil || p.ValidateURL.String() == "" {
```
Can we just set this a few lines above in `p.setProviderDefaults(providerDefaults{...` where the `validateURL` is currently set to `nil`?

codablock replied:
Done

codablock replied:
I had to remove this change again as it made `TestNewAzureProvider` fail now. Looking at the code, I realized that we can't simply rely on `p.setProviderDefaults` here, because the default value is dependent on `p.ProfileURL` being defaulted or not.

NickMeves commented on providers/azure.go:
```diff
@@ -210,3 +270,8 @@ func (p *AzureProvider) GetEmailAddress(ctx context.Context, s *sessions.Session
 }
+
+// ValidateSessionState validates the AccessToken
+func (p *AzureProvider) ValidateSessionState(ctx context.Context, s *sessions.SessionState) bool {
```
Nice! Thanks for this addition.

codablock replied:
Credits must go to @m-wcislo who did the original PR :)

NickMeves reviewed (changes requested):
Overall this looks good!

My biggest concern is the hardcoded 10 minute time shift to attempt to refresh early.

codablock commented:
Rebased on master and handled the requested changes (except the NIT as I'm for sticking with surrounding style and doing style fixes in separate PRs).

NickMeves reviewed (approved):
Looks good! Thanks for aligning the refresh logic to other providers (even if it does have weaknesses that we need to refactor as a whole 😅 ).

Please add a CHANGELOG entry.  I think this one will need both an `Important Notes` detailing the new access token refresh logic (and a best practice note to set the `cookie-refresh` setting to the Access Token lifespan) & a `Breaking Changes` detailing the `ValidateSessionState` security improvement (environments that had cookie-session TTLs longer than Azure Access/ID Token TTLs might be impacted by this change).

I'll look to merge this once we confirm the next release is v7 (major)

codablock commented:
@NickMeves please see https://github.com/oauth2-proxy/oauth2-proxy/pull/787 for CHANGELOG.

codablock commented:
@NickMeves I've rebased on develop, fixed merge conflicts and squashed everything into the first commit. I've also added CHANGELOG entries to this PR.

JoelSpeed commented on CHANGELOG.md:
```diff
@@ -19,9 +19,16 @@
   via the login url. If this option was used in the past, behavior will change with this release as it will
   affect the tokens returned by Azure. In the past, the tokens were always for `https://graph.microsoft.com` (the default)
   and will now be for the configured resource (if it exists, otherwise it will run into errors)
+- [#754](https://github.com/oauth2-proxy/oauth2-proxy/pull/754) The Azure provider now has token refresh functionality implemented. This means that there won't
+  be any redirects in the browser anymore when tokens expire, but instead a token refresh is initiated
+  in the background, which leads to new tokens being returned in the cookies.
+- [#754](https://github.com/oauth2-proxy/oauth2-proxy/pull/754) Pleas note that `--cookie-refresh` must be 0 (the default) or equal to the token lifespan configured in Azure AD to make
+  Azure token refresh reliable. Setting this value to 0 means that it relies on the provider implementation
+  to decide if a refresh is required.
```
This can be a sub-bullet
```suggestion
- [#754](https://github.com/oauth2-proxy/oauth2-proxy/pull/754) The Azure provider now has token refresh functionality implemented. This means that there won't
  be any redirects in the browser anymore when tokens expire, but instead a token refresh is initiated
  in the background, which leads to new tokens being returned in the cookies.
  - Please note that `--cookie-refresh` must be 0 (the default) or equal to the token lifespan configured in Azure AD to make
  Azure token refresh reliable. Setting this value to 0 means that it relies on the provider implementation
  to decide if a refresh is required.
```

JoelSpeed commented on providers/azure.go:
```diff
@@ -151,6 +157,60 @@ func (p *AzureProvider) Redeem(ctx context.Context, redirectURL, code string) (s
+
+	origExpiration := s.ExpiresOn
+
+	err := p.redeemRefreshToken(ctx, s)
+
+	if err != nil {
```
Nit: Please pair error declaration and checks, this is how they normally are in go 
```suggestion
	err := p.redeemRefreshToken(ctx, s)
	if err != nil {
```

JoelSpeed commented on providers/azure.go:
```diff
@@ -151,6 +157,60 @@ func (p *AzureProvider) Redeem(ctx context.Context, redirectURL, code string) (s
+	s.IDToken = jsonResponse.IDToken
+	s.RefreshToken = jsonResponse.RefreshToken
+	s.ExpiresOn = &expires
+	return
```
When a session is refreshed, you should set the `CreatedAt` to `time.Now()` else the proxy will spew lots of refreshing x old session messages

knightian commented:
Wen merge 😅

NickMeves commented:
@codablock I think once you resolve the couple of notes @JoelSpeed left this is good to merge.

FYI - There is a draft in progress of the refresh timing logic centralization we discussed: https://github.com/oauth2-proxy/oauth2-proxy/pull/886

codablock commented:
Sorry for the delay, I didn't notice the comments from JoelSpeed. I've applied suggestions/fixes and rebased on master. I hope we're ready for the merge now :)

knightian commented:
ping ping ping

NickMeves commented:
> ping ping ping 😅😃

You can always Fork if we are too slow for you 😛

knightian commented:
> > ping ping ping 😅😃
> 
> You can always Fork if we are too slow for you 😛

Haha it is true what you say, and I thought about it, but because it's a security thing I decided I'll stay on official-master

NickMeves reviewed (approved):
LGTM!

knightian commented:
woop W00P!
